akka-stream-tests/src/test/scala/akka/stream/scaladsl/RestartSpec.scala

```diff
@@ -0,0 +1,461 @@
+package akka.stream.scaladsl
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.Done
+import akka.stream.ActorMaterializer
+import akka.stream.testkit.StreamSpec
+import akka.stream.testkit.scaladsl.{ TestSink, TestSource }
+import akka.testkit.DefaultTimeout
+
+import scala.concurrent.Promise
+import scala.concurrent.duration._
+import scala.util.{ Failure, Success }
+
+class RestartSpec extends StreamSpec with DefaultTimeout {
+
+  implicit val mat = ActorMaterializer()
+  import system.dispatcher
+
+  "A restart with backoff source" should {
+    "run normally" in {
+      val created = new AtomicInteger()
+      val probe = RestartSource.withBackoff(10.millis, 20.millis, 0) { () ⇒
+        created.incrementAndGet()
+        Source.repeat("a")
+      }.runWith(TestSink.probe)
+
+      probe.requestNext("a")
+      probe.requestNext("a")
+      probe.requestNext("a")
+      probe.requestNext("a")
+      probe.requestNext("a")
+
+      created.get() should ===(1)
+    }
+
+    "restart on completion" in {
+      val created = new AtomicInteger()
+      val probe = RestartSource.withBackoff(10.millis, 20.millis, 0) { () ⇒
+        created.incrementAndGet()
+        Source(List("a", "b"))
+      }.runWith(TestSink.probe)
+
+      probe.requestNext("a")
+      probe.requestNext("b")
+      probe.requestNext("a")
+      probe.requestNext("b")
+      probe.requestNext("a")
+
+      created.get() should ===(3)
```

Comment:
Is it what you'd expect from Akka's backoff supervisor? Because it's exactly the same the behavior, and has exactly the same use cases. I'd like to use this feature for running things like CQRS event streams, Kafka subscriptions, and broadcasting ConductR event streams. You can't just stop processing all CQRS events in your system just because the source producing the stops. The source stopping (depending on the source) may indicate that the server that produced it is shutting down (perhaps due to an upgrade) and has cleanly closed its connections. An upgrade shouldn't cause me to stop doing what I need to do, I need to reestablish that connection (eg, if it's a rolling upgrade that caused the stream to stop, then after I get disconnected I want to connect to another node).